bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,963 @@
+	}
+
+	b := credentialsBundle{
+		channel: channelCreds,
+		call:    clc,
+	}
+
+	d := grpc.WithCredentialsBundle(b)
+	opts = append(opts, d)
+
+	return opts, nil
+}
+
+// getChannelCredentials extracts the channel credentials (credentials for use)
+// with all calls on this client.
+func getChannelCredentials(credsProto *pb.ChannelCredential, sslTargetName string) (credentials.TransportCredentials, error) {
+	var creds credentials.TransportCredentials
+	v := credsProto.GetValue()
+	switch t := v.(type) {
+	case *pb.ChannelCredential_Ssl:
+		pem := t.Ssl.GetPemRootCerts()
+
+		cert, err := x509.ParseCertificate([]byte(pem))
+		if err != nil {
+			return nil, err
+		}
+
+		pool := x509.NewCertPool()
+		pool.AddCert(cert)
+
+		creds = credentials.NewClientTLSFromCert(pool, sslTargetName)
+		if err != nil {
+			return nil, err
+		}
+	case *pb.ChannelCredential_None:
+		creds = insecure.NewCredentials()
+	default:
+		ctx := context.Background()
+		c, err := gauth.FindDefaultCredentials(ctx, []string{"https://www.googleapis.com/auth/cloud-platform"}...)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO(developer): Determine how to pass this call option back to caller
+		option.WithTokenSource(c.TokenSource)
+
+		return nil, nil
+	}
+	return creds, nil
+}
+
+// goTestProxyServer represents an instance of the test proxy server. It keeps
+// a reference to individual clients instances (stored in a testClient object).
+type goTestProxyServer struct {
+	pb.UnimplementedCloudBigtableV2TestProxyServer
+	clientIDs map[string]testClient // clientIDs has all of the bigtable.Client objects under test
+}
+
+// CreateClient responds to the CreateClient RPC. This method adds a new client
+// instance to the goTestProxyServer
+func (s *goTestProxyServer) CreateClient(ctx context.Context, req *pb.CreateClientRequest) (*pb.CreateClientResponse, error) {
+	if req.ClientId == "" ||
+		req.DataTarget == "" ||
+		req.ProjectId == "" ||
+		req.InstanceId == "" {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s must provide ClientId, DataTarget, ProjectId, and InstanceId", logLabel))
+	}
+
+	if _, exists := s.clientIDs[req.ClientId]; exists {
+		return nil, stat.Error(codes.AlreadyExists,
+			fmt.Sprintf("%s: ClientID already exists", logLabel))
+	}
+
+	opts, err := getCredentialsOptions(req)
+	if err != nil {
+		return nil, stat.Error(codes.Unauthenticated,
+			fmt.Sprintf("%s: failed to get credentials: %v", logLabel, err))
+	}
+
+	conn, err := grpc.Dial(req.DataTarget, opts...)
+	if err != nil {
+		return nil, stat.Error(codes.Unknown, fmt.Sprintf("%s: failed to create connection: %v", logLabel, err))
+	}
+
+	localCtx, cancel := context.WithCancel(context.Background())
+	c, err := bigtable.NewClient(localCtx, req.ProjectId, req.InstanceId, option.WithGRPCConn(conn))
+	if err != nil {
+		cancel()
+		return nil, stat.Error(codes.Internal,
+			fmt.Sprintf("%s: failed to create client: %v", logLabel, err))
+	}
+
+	if s.clientIDs == nil {
```

Comment:
Does this need to be threadsafe?

Reply:
Done.

Reply:
This nil check isn't theadsafe still. See the comment above about wrapping the method in a lock.

Reply:
Done (I hope!)